front/pages/api/w/[wId]/vaults/[vId]/data_sources/index.ts

```diff
@@ -0,0 +1,633 @@
+import type {
+  DataSourceType,
+  DataSourceViewType,
+  PlanType,
+  UserType,
+  WithAPIErrorResponse,
+  WorkspaceType,
+} from "@dust-tt/types";
+import {
+  CONNECTOR_PROVIDERS,
+  ConnectorConfigurationTypeSchema,
+  ConnectorsAPI,
+  CoreAPI,
+  DEFAULT_EMBEDDING_PROVIDER_ID,
+  DEFAULT_QDRANT_CLUSTER,
+  dustManagedCredentials,
+  EMBEDDING_CONFIGS,
+  ioTsParsePayload,
+  isDataSourceNameValid,
+  sendUserOperationMessage,
+  WebCrawlerConfigurationTypeSchema,
+} from "@dust-tt/types";
+import { isLeft } from "fp-ts/lib/Either";
+import * as t from "io-ts";
+import * as reporter from "io-ts-reporters";
+import type { NextApiRequest, NextApiResponse } from "next";
+
+import config from "@app/lib/api/config";
+import { getDataSource } from "@app/lib/api/data_sources";
+import { withSessionAuthenticationForWorkspace } from "@app/lib/api/wrappers";
+import type { Authenticator } from "@app/lib/auth";
+import { getOrCreateSystemApiKey } from "@app/lib/auth";
+import {
+  getDefaultDataSourceDescription,
+  getDefaultDataSourceName,
+  isConnectionIdRequiredForProvider,
+  isConnectorProviderAllowedForPlan,
+  isConnectorProviderAssistantDefaultSelected,
+  isValidConnectorSuffix,
+} from "@app/lib/connector_providers";
+import { DataSourceResource } from "@app/lib/resources/data_source_resource";
+import { DataSourceViewResource } from "@app/lib/resources/data_source_view_resource";
+import { VaultResource } from "@app/lib/resources/vault_resource";
+import { ServerSideTracking } from "@app/lib/tracking/server";
+import { isDisposableEmailDomain } from "@app/lib/utils/disposable_email_domains";
+import logger from "@app/logger/logger";
+import { apiError } from "@app/logger/withlogging";
+
+export const PostDataSourceWithProviderRequestBodySchema = t.type({
+  provider: getConnectorProviderCodec(),
+  name: t.union([t.string, t.undefined]),
+  connectionId: t.union([t.string, t.undefined]),
+  configuration: ConnectorConfigurationTypeSchema,
+});
+// Sorcery: Create a union type with at least two elements to satisfy t.union
+export function getConnectorProviderCodec(): t.Mixed {
```

Comment:
Seems surprising that this does not exist out of the box?

Reply:
Yes I lost some time on it, and eventually got something that works and after I realized we were using this kind of sorcery elsewhere in the codebase, so I went with it!